vm/devices/storage/storvsc_driver/src/lib.rs

```diff
@@ -0,0 +1,920 @@
+        &mut self,
+        writer: &mut queue::WriteHalf<'_, M>,
+        operation: storvsp_protocol::Operation,
+        status: storvsp_protocol::NtStatus,
+        transaction_id: u64,
+        payload: &P,
+    ) -> Result<(), StorvscError> {
+        let payload_bytes = payload.as_bytes();
+        self.send_vmbus_packet(
+            &mut writer.batched(),
+            OutgoingPacketType::InBandWithCompletion,
+            transaction_id,
+            operation,
+            status,
+            payload_bytes,
+        )?;
+        Ok(())
+    }
+
+    /// Send a GPA Direct packet over VMBus.
+    fn send_gpa_direct_packet<M: RingMem, P: IntoBytes + Immutable + KnownLayout>(
+        &mut self,
+        writer: &mut queue::WriteHalf<'_, M>,
+        operation: storvsp_protocol::Operation,
+        status: storvsp_protocol::NtStatus,
+        transaction_id: u64,
+        payload: &P,
+        gpa_start: u64,
+        byte_len: usize,
+    ) -> Result<(), StorvscError> {
+        let payload_bytes = payload.as_bytes();
+        let start_page: u64 = gpa_start / PAGE_SIZE as u64;
+        let end_page: u64 = (gpa_start + (byte_len + PAGE_SIZE - 1) as u64) / PAGE_SIZE as u64;
+        let gpas: Vec<u64> = (start_page..end_page).collect();
+        let pages =
+            PagedRange::new(gpa_start as usize % PAGE_SIZE, byte_len, gpas.as_slice()).unwrap();
+        self.send_vmbus_packet(
+            &mut writer.batched(),
+            OutgoingPacketType::GpaDirect(&[pages]),
+            transaction_id,
+            operation,
+            status,
+            payload_bytes,
+        )?;
+        Ok(())
+    }
+
+    /// Send a VMBus packet.
+    fn send_vmbus_packet<M: RingMem>(
+        &mut self,
+        writer: &mut queue::WriteBatch<'_, M>,
+        packet_type: OutgoingPacketType<'_>,
+        transaction_id: u64,
+        operation: storvsp_protocol::Operation,
+        status: storvsp_protocol::NtStatus,
+        payload: &[u8],
+    ) -> Result<(), StorvscError> {
+        let header = storvsp_protocol::Packet {
+            operation,
+            flags: 0,
+            status,
+        };
+
+        // storvsp limits the size of the completion packet to the size of the request packet,
+        // so we need to pad the payload to the maximum size to ensure we get a complete response.
+        // The maximum size includes header + payload.
+        let len = size_of_val(&header) + size_of_val(payload);
+        let padding = [0; storvsp_protocol::SCSI_REQUEST_LEN_MAX];
+        let (payload_bytes, padding_bytes) = if len > storvsp_protocol::SCSI_REQUEST_LEN_MAX {
+            (
+                &payload[..storvsp_protocol::SCSI_REQUEST_LEN_MAX - size_of_val(&header)],
+                &[][..],
+            )
+        } else {
+            (
+                payload,
+                &padding[..storvsp_protocol::SCSI_REQUEST_LEN_MAX - len],
+            )
+        };
```

Comment:
I would have expected us to just panic if the vmbus packet is bigger than what we expect for our protocol, since we're truncating it anyways in a way that is potentially semantically unsafe. But I imagine there are some quirks in the vstor protocol ( :) ), so what quirks are we hitting here that require the code to hande this?

Reply:
The storvsp implementation also does this unusual truncation: https://github.com/microsoft/openvmm/blob/3c2dda8b996f4425e74a8a961d52f4f6c90b9f62/vm/devices/storage/storvsp/src/lib.rs#L457-L471

Reply:
I have looked over this code multiple times, including the comments in storvsp. The comments linked from storvsp talk about padding or truncating the *payload* specifically, but it seems to *also* consider the length of the header when padding/truncating. The size of a SCSI request (the payload) is 52 bytes, which is the length limited to in storvsp and here. Therefore, I believe the intention is to pad/truncate the payload only. I have modified the code here in storvsc (not in storvsp) to remove truncation (to avoid any hidden issues) and to pad *only* the payload. It is worth revisiting the code in storvsp to see if this is an error there.

Reply:
Okay, I think that's reasonable. Please ensure that this is covered in a test case, and then I'm happy to consider this resolved.